main.cpp

```diff
@@ -105,6 +106,21 @@ struct FloatType
 
     operator float() const { return *a; }
 
+    FloatType& apply( std::function< FloatType&( float& )> func )
+    {
+        if (func != nullptr )
+            func( *a );
```

Comment:
`func()` returns `FloatType&`, so you should return that object.   If you mentally step through the execution, you'll see that the returned object is `*this`.
- [ ] revise.